pkg/labeller/labeller.go

```diff
@@ -16,39 +15,79 @@ type FetchableObject struct {
 }
 
 type Labeller struct {
-	APIProxy       apiproxy.APIProxy
-	ObjectMeta     *metav1.ObjectMeta
-	ComponentLabel string
-	ComponentName  string
-	Annotations    map[string]string
+	ObjectMeta                 *metav1.ObjectMeta
+	ComponentObjectsNamePrefix string
+	ComponentFullName          string
+	annotations                map[string]string
+
+	instanceName string
+
+	objectsNamePrefixBase string
+}
+
+func NewLabellerForGlobalComponent(objectMeta *metav1.ObjectMeta, name consts.ComponentType, objectsNamePrefix string, annotations map[string]string) Labeller {
+	l := Labeller{
+		ObjectMeta: objectMeta,
+
+		ComponentFullName:          string(name),
+		ComponentObjectsNamePrefix: objectsNamePrefix,
+
+		objectsNamePrefixBase: objectsNamePrefix,
+
+		annotations: annotations,
+	}
+
+	return l
+}
+
+func NewLabellerForComponentInstance(objectMeta *metav1.ObjectMeta, nameBase consts.ComponentType, objectsNamePrefixBase, instanceName string, annotations map[string]string) Labeller {
+	l := Labeller{
+		ObjectMeta: objectMeta,
+
+		ComponentFullName:          string(nameBase),
+		ComponentObjectsNamePrefix: objectsNamePrefixBase,
+
+		instanceName: instanceName,
+
+		objectsNamePrefixBase: objectsNamePrefixBase,
+
+		annotations: annotations,
+	}
+
+	if instanceName != "" && instanceName != consts.DefaultName {
+		l.ComponentFullName = fmt.Sprintf("%s-%s", nameBase, instanceName)
+		l.ComponentObjectsNamePrefix = fmt.Sprintf("%s-%s", objectsNamePrefixBase, instanceName)
+	}
+
+	return l
```

Comment:
Not sure if some style guide mentions it, but I think it is good practice to write constructors like that:
```go
func NewLabeller(...) Labeller {
    // variables and ifs
    return Labeller{
       ...
    }
}
```
than creating an instance, and redefining its fields later.